Improves PyQuil folding functions and tests

andreamari opened the pull request:
1. Partially solves the issue #9, improving the creation of empty circuits retaining the input properties.

2. Implements several new tests (unitary equivalence, mutation of input and output properties)

willzeng commented on mitiq/folding_pyquil.py:
```diff
@@ -41,7 +41,7 @@ def fold_gates(circuit: Program, sub_indices: List[int]) -> Program:
     determined by sub_indices."""
 
-    out = Program()
+    out = circuit.copy_everything_except_instructions()
```
neat find! I didn't know they already had that function :)

rmlarose replied:
Also wasn't aware of this -- nice!

willzeng commented on mitiq/tests/test_folding_pyquil.py:
```diff
@@ -1,48 +1,112 @@
+
+# NUM_SHOTS must be 1 in order to pass the next tests.
+# This is due to the fact that num_shots are not retained by
+# some methods of the Program class (bug?)
```
Can you say a little more about what this means? What is an example method that loses the num_shots?

andreamari replied:
This part is changed. That comment is removed.

andreamari replied:
This part is changed. That comment is removed.

willzeng commented on mitiq/tests/test_folding_pyquil.py:
```diff
@@ -1,48 +1,112 @@
         out = unitary_folding(circ, c)
         actual_c = len(out) / len(circ)
-        print("Input stretch: {:}    Real stretch: {}".format(c, actual_c))
+        # test lenght scaling
```
```suggestion
        # test length scaling
```

willzeng commented on mitiq/tests/test_folding_pyquil.py:
```diff
@@ -1,48 +1,112 @@
         out = local_folding(circ, c, stretcher=sampling_stretcher)
         actual_c = len(out) / len(circ)
-        print("Input stretch: {:}    Real stretch: {}".format(c, actual_c))
+        # test lenght scaling
```
```suggestion
        # test length scaling
```

willzeng commented on mitiq/tests/test_folding_pyquil.py:
```diff
@@ -1,48 +1,112 @@
+from copy import deepcopy
+
+
+STRETCH_VALS = [1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0]
```
For testing purposes might be good to change this to use not only integer and half-integer values and not to be evenly spaced, e.g.

```suggestion
STRETCH_VALS = [1.0, 1.3, 2.0, 2.6, 3.0, 3.5, 4.0]
```

willzeng commented on mitiq/tests/test_folding_pyquil.py:
```diff
@@ -1,48 +1,112 @@
+        u_in = program_unitary(circ, 2)
+        u_out = program_unitary(out, 2)
+        assert (u_in == u_out).all
+        # test input is not mutated up to "_synthesized_instructions"
```
Given that this set of testing instructions is copied between all three tests, would it be better to have this as a separate function that is then called three times?

rmlarose replied:
+1

andreamari replied:
Yes!

andreamari commented:
> Looks good @andreamari! I made mostly cosmetic comments.
> 
> One thing I would like to understand better is what is going on with the num_shots stuff. That might be a bigger problem than we realize. Could you say a little more about it?

The problem with `num_shots` and similar properties of a `Program` is that they are not preserved by the main methods (e.g. `__add__` and `__getitem__` ) the `Program` class.  I think it is better if we discuss about this problem in the dedicated issue #9.

willzeng commented:
Your right @andreamari am continuing this discussion on #9 I should have checked there beforehand

rmlarose commented on mitiq/tests/test_folding_pyquil.py:
```diff
@@ -1,48 +1,112 @@
 import random
 from pyquil import Program
-from pyquil.gates import X, Y, Z
-from mitiq.folding_pyquil import local_folding, unitary_folding, sampling_stretcher, left_stretcher
```
More of a future point, but I think it would be good to discuss common naming conventions for (folding) functions. I generally default to function names which are verbs that (hopefully) describe what the function does, e.g. `fold_gates_from_left`, but there are different styles. Eventually we'll want consistency in folding function names, but again more of a future point.

willzeng replied:
+1 to using verbs for function names. I like that convention.

andreamari replied:
Agree. I will use the same names used for Cirq

rmlarose commented on mitiq/tests/test_folding_pyquil.py:
```diff
@@ -1,48 +1,112 @@
     for _ in range(depth):
-        prog += random.choice([X(0), Y(0), Z(0)])
+        prog += random.choice(gate_set)
+    prog.num_shots = NUM_SHOTS
```
Should this be `prog.wrap_in_numshots_loop`? Is there a difference?

andreamari replied:
In practice it's the same. But I used your suggestion because `wrap_in_numshots_loop` looks more similar to what a user would do in a real scenario.

andreamari commented:
I think I have applied all suggestions and solved the main open problems of PyQuil folding functions:
 
- Better names (and docs) for functions, which are now consistent with those of `folding_cirq.py`.
- Added the possibility of using a seed, in the same way as we do for Cirq.
- Fixed issue #9 (circuit properties not retained) by properly appending instructions to the output circuit.
- Removed repetitions of the same code in tests with the new function `apply_folding_tests`.
- New test for seed parameter.

rmlarose commented on mitiq/folding_pyquil.py:
```diff
@@ -1,76 +1,110 @@
+
+    ngates = len(circuit)
+    num_to_fold = int(ngates * (stretch - 1) / 2)
+    sub_indices = np.random.choice(range(ngates), num_to_fold, replace = False)
```
```suggestion
    sub_indices = np.random.choice(range(ngates), num_to_fold, replace=False)
```

rmlarose replied:
(PEP8)

andreamari replied:
Right, thanks!

rmlarose commented on mitiq/folding_pyquil.py:
```diff
@@ -1,76 +1,110 @@
+def fold_local(
+        circuit: Program,
+        stretch: float,
+        fold_method: Callable = fold_gates_from_left,
```
```suggestion
        fold_method: Callable[[Callable, float, Tuple[Any]], Circuit] = fold_gates_from_left,
```

rmlarose replied:
^ Add type signature to `fold_method` arg.

andreamari replied:
Ok. But isn't this the correct signature: `Callable[[Program, float, Tuple[Any]], Program]` ?

willzeng replied:
yep that looks right @andreamari

willzeng commented:
LGTM once that last type signature is in place @andreamari
